File: Heap/MinHeap.java
import java.util.ArrayList;
import java.util.Collections;


class Heap {
    private List<Integer> data = new ArrayList<>();
    
    public int size() {
        return data.size;
    }

    public int peek() {
        return data.get(0); // data (0]
    }

    public void push(int ele) {
        // 1. Complete Binary Tree
        data.add (ele); // 0(1)

        // 2. Heap Order Property
        upheapify(data.size() - 1);
    }

    private void upheapify(int idx) {
        if(idx == 0) return;
        int parent = (idx - 1) / 2;
        
        if(data.get(parent) > data.get (idx)) {
            Collections. swap(data, idx, parent);        
            upheapify (parent) ;
        }

    }

    public void pop() {
        // 1. Complete Binary Tree
        int n = data.size();
        Collections.swap(data, 0, n - 1); // Swap root with last element 
        data.remove(n -1);

        // 2. Heap Order Property
        downheapify(0);
    }
    
    private voide downheapify(int idx) {
        int min = idx;
        int l = 2 * idx + 1, r = 2 * idx + 2;

        // LEFT
        if(l < data.size() && data.get(l) < data.get(min)) {
            min = l;
        }

        // RIGHT
        if(r < data.size() && data.get(r) < data.get(min)) {
            min = r;
        }

        if(min == l) {
            Collections.swap(data, idx, l);
            downheapify(l);
        } else if(min == r) {
            Collections.swap(data, idx, r);
            downheapify(r);
        }
    }

    public static void main(String[] args) {
        Heap heap = new Heap();
        heap.push(10);
        heap.push(20);
        heap.push(5);
        heap.push(15);
        
        System.out.println("Heap size: " + heap.size());
        System.out.println("Top element: " + heap.peek());
        
        heap.pop();
        System.out.println("Top element after pop: " + heap.peek());
    }
}

File: Heap/MinHeap.js
class Heap {
  constructor(){
    this.heap = [];
  }
  getParentIdx(idx){
    return Math.floor((idx - 1) / 2);
  }
  getLeftChildIdx(idx){
    return 2 * idx + 1;
  }
  getRightChildIdx(idx){
    return 2 * idx + 2;
  }
  swap(ith, jth){
    [this.heap[ith], this.heap[jth]] = [this.heap[jth], this.heap[ith]];
  }
  insert(val){
    this.heap.push(val);
    this.upheapify();
  }
  upheapify(){
    let idx = this.heap.length -1;
    while(idx > 0 && this.heap[this.getParentIdx(idx)] > this.heap[idx]){
      this.swap(idx, this.getParentIdx(idx));
      idx = this.getParentIdx(idx);
    }
  }
  remove(){
    if(this.heap.length === 0) return null;
    if(this.heap.length === 1) return this.heap.pop();

    const root = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.downheapify();
    
    return root;
  }
  downheapify(){
    let idx = 0;
    while (this.getLeftChildIdx(idx) < this.heap.length) {
      let smallerChildIndex = this.getLeftChildIdx(idx);
      let rightChildIndex = this.getRightChildIdx(idx);
      if( rightChildIndex < this.heap.length && this.heap[rightChildIndex] < this.heap[smallerChildIndex]){
        smallerChildIndex = rightChildIndex;
      }
      if (this.heap[idx] <= this.heap[smallerChildIndex]) break;
      this.swap(idx, smallerChildIndex);
      idx = smallerChildIndex;
    }
  }
  peek() {
    return this.heap[0];
  }

  size() {
    return this.heap.length;
  }
}